### ISSUE TITLE
Smart Playlists (iTunes-style)

### ISSUE THREAD

**Issue by GlassedSilver** [Tzahi12345/YoutubeDL-Material#126]
So I just wanted to create a playlist and something came to my mind...

What if we had Smart Playlists?

One of the reasons why I'm so hesitant to move away from iTunes at least for music is smart playlists, despite being an Android-first smartphone user now (over previously my iPhone, oh and I mean just the smartphone hemisphere. I'm still "desktop-first" if we're looking at computing in general... But I digress...).

Well anyhow, iTunes isn't the topic. However it is a shining example of how Smart Playlists have to be done: touch on (almost?) any possible value there is, so you can set filters like "unwatched in x months AND by this artist OR featured in this collection"...

Really you can get super crazy there by combining, nesting, OR-ing anything you like.

I would love to create playlists like "4K content" or "Unwatched AND category:news" or something along those lines. Smart Playlists are an easily customizable way to empower the viewing habits of a user and could perfectly _compliment_ a recommendation system.

So "AI"-suggested and user-defined lists alongside could actually serve a user better suggestions than the YT algorithm (that to me... works pretty great some weeks, other weeks it just seems to love to shove one kind of videos and only that one haha (probs some trending-push))

PS: When I list examples like "OR" or parameters like "unwatched" I mean those as inspirations. Don't write them down, just scratching the surface here... :D 
PPS: Sorry if that sounded demanding, it's just that I have a very deep running passion for smart playlists and one of the things I miss most about Apple in my everyday life is that much of their software had/has a strong emphasis on stuff like smart playlists/smart folders/dynamically generated searches, etc... (latter of which is luckily a thing on Windows as well :) )
PPS: Easy to imagine I love dynamic collections in Steam as well, right? Coming full circle here to my love for Steam's new UI with the hovering animation. :P

**Comment by Tzahi12345** [comment]
Good idea! But this is probably something for 4.3 and beyond (like you hinted in your PPS). It doesn't sound demanding at all In fact, I really like these "long shot" ideas, because it helps create a mental timeline for the future of the app. It's not just "this is where we are, now what do we add?", we at least have an idea of the direction. Best of all, other users can add/improve the suggestion as well.

Smart playlists depends on having the category/tags idea, as well as additional metadata for more powerful boolean expressions (unwatched, for example). So once those are implemented, this can move along.

There's still some stuff to consider, like when are the downloaded videos checked to see whether they fit into the smart playlist? Here's what I think:
* When a new video is downloaded, it checks to see if it fits into the smart playlist
* When a new smart playlist is created, all existing videos are checked to see if they apply to it
  * If you are able to change the rules for a smart playlist *after* creation, then after updating the rules, all existing videos need to be re-checked against the new rules

In general, great idea! You're right in that this will sit in the back burner for a bit :) As always, thanks for the suggestion, always appreciated.

**Comment by web-connect** [comment]
FWIW what I do is I have youtubedl-material be my media downloader, and then I load up plex and it manages my playlists. It has those features, with auto playlists and auto categories etc. 

I like the fact that Plex is dedicated to making my user experience of playback really good, and this project focused on being really good and syncing/pulling content down. 

But this is a really interesting concept!

**Comment by GlassedSilver** [comment]
I'm a Plex user myself, but the issue is that Plex is unfit in many ways. If you don't add episodes in the exact order the got published or as part of a playlist then thumbnails become a real issue.

Also youtubedl-material can be a really good solution for content of the web-originated media, including and presenting properly the auxiliary information like special meta data, and video recommendations and more.

If you use movies as library type it's probably better in Plex, but then you're missing out on other things again and whilst I started off wishing to make Youtubedl-material work along Plex I have since started believing that whilst having all my media in one place is nice specialized tools are sometimes better than the Swiss army knife approach where one tool can do all but lacks in some special tasks a special tool can provide a better experience.

That being said no matter if you export your data (which should always be easy - hence my passion for storing thumbnails and the info json in separate files instead of just youtubedl-material's internal database!) or keep it in youtubedl-material you will always need the downloader to be solid.

A good exporting experience is also one of the reasons why I advocate the improvements in path flexibility!
